In [1]:
import numpy as np
import pandas as pd

In [2]:
cars=pd.read_csv('CarPrice_Assignment.csv')

In [3]:
cars.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

## Understanding the data

In [4]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

 ## Label Encoding

In [5]:
from sklearn.preprocessing import LabelEncoder

In [6]:
le= LabelEncoder()

In [7]:
# For CarName column

In [8]:
cars["CarName"]=le.fit_transform(cars["CarName"])

In [9]:
cars.head()     # The column has been encoded

car_ID  symboling  CarName fueltype aspiration doornumber      carbody  \
0       1          3        2      gas        std        two  convertible   
1       2          3        3      gas        std        two  convertible   
2       3          1        1      gas        std        two    hatchback   
3       4          2        4      gas        std       four        sedan   
4       5          2        5      gas        std       four        sedan   

  drivewheel enginelocation  wheelbase  ...  enginesize  fuelsystem  \
0        rwd          front       88.6  ...         130        mpfi   
1        rwd          front       88.6  ...         130        mpfi   
2        rwd          front       94.5  ...         152        mpfi   
3        fwd          front       99.8  ...         109        mpfi   
4        4wd          front       99.4  ...         136        mpfi   

   boreratio  stroke compressionratio horsepower  peakrpm citympg  highwaympg  \
0       3.47    2.68              9.0        111     5000      21          27   
1       3.47    2.68              9.0        111     5000      21          27   
2       2.68    3.47              9.0        154     5000      19          26   
3       3.19    3.40             10.0        102     5500      24          30   
4       3.19    3.40              8.0        115     5500      18          22   

     price  
0  13495.0  
1  16500.0  
2  16500.0  
3  13950.0  
4  17450.0  

[5 rows x 26 columns]

In [10]:
# Mapping the encoded values with original data

In [11]:
carnamemapping={idx:label for idx,label in enumerate(le.classes_)}

In [12]:
print(carnamemapping)

{0: 'Nissan versa', 1: 'alfa-romero Quadrifoglio', 2: 'alfa-romero giulia', 3: 'alfa-romero stelvio', 4: 'audi 100 ls', 5: 'audi 100ls', 6: 'audi 4000', 7: 'audi 5000', 8: 'audi 5000s (diesel)', 9: 'audi fox', 10: 'bmw 320i', 11: 'bmw x1', 12: 'bmw x3', 13: 'bmw x4', 14: 'bmw x5', 15: 'bmw z4', 16: 'buick century', 17: 'buick century luxus (sw)', 18: 'buick century special', 19: 'buick electra 225 custom', 20: 'buick opel isuzu deluxe', 21: 'buick regal sport coupe (turbo)', 22: 'buick skyhawk', 23: 'buick skylark', 24: 'chevrolet impala', 25: 'chevrolet monte carlo', 26: 'chevrolet vega 2300', 27: 'dodge challenger se', 28: 'dodge colt (sw)', 29: 'dodge colt hardtop', 30: 'dodge coronet custom', 31: 'dodge coronet custom (sw)', 32: 'dodge d200', 33: 'dodge dart custom', 34: 'dodge monaco (sw)', 35: 'dodge rampage', 36: 'honda accord', 37: 'honda accord cvcc', 38: 'honda accord lx', 39: 'honda civic', 40: 'honda civic (auto)', 41: 'honda civic 1300', 42: 'honda civic 1500 gl', 43: 'hon

In [13]:
# Now encoding all the columns whose data is present in string format

In [14]:
cols = ["fueltype","aspiration","doornumber","carbody","drivewheel","enginelocation","enginetype","fuelsystem","cylindernumber"]

In [15]:
for col in cols:
    cars[col] = le.fit_transform(cars[col])
    colMapping = {idx:label for idx,label in enumerate(le.classes_)}
    print(colMapping)

{0: 'diesel', 1: 'gas'}
{0: 'std', 1: 'turbo'}
{0: 'four', 1: 'two'}
{0: 'convertible', 1: 'hardtop', 2: 'hatchback', 3: 'sedan', 4: 'wagon'}
{0: '4wd', 1: 'fwd', 2: 'rwd'}
{0: 'front', 1: 'rear'}
{0: 'dohc', 1: 'dohcv', 2: 'l', 3: 'ohc', 4: 'ohcf', 5: 'ohcv', 6: 'rotor'}
{0: '1bbl', 1: '2bbl', 2: '4bbl', 3: 'idi', 4: 'mfi', 5: 'mpfi', 6: 'spdi', 7: 'spfi'}
{0: 'eight', 1: 'five', 2: 'four', 3: 'six', 4: 'three', 5: 'twelve', 6: 'two'}


In [16]:
cars.head()      # All the columns are encoded

car_ID  symboling  CarName  fueltype  aspiration  doornumber  carbody  \
0       1          3        2         1           0           1        0   
1       2          3        3         1           0           1        0   
2       3          1        1         1           0           1        2   
3       4          2        4         1           0           0        3   
4       5          2        5         1           0           0        3   

   drivewheel  enginelocation  wheelbase  ...  enginesize  fuelsystem  \
0           2               0       88.6  ...         130           5   
1           2               0       88.6  ...         130           5   
2           2               0       94.5  ...         152           5   
3           1               0       99.8  ...         109           5   
4           0               0       99.4  ...         136           5   

   boreratio  stroke  compressionratio  horsepower  peakrpm  citympg  \
0       3.47    2.68               9.0         111     5000       21   
1       3.47    2.68               9.0         111     5000       21   
2       2.68    3.47               9.0         154     5000       19   
3       3.19    3.40              10.0         102     5500       24   
4       3.19    3.40               8.0         115     5500       18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

## Separating input and output columns

In [17]:
# output 
y=cars["price"]
y.head()

0    13495.0
1    16500.0
2    16500.0
3    13950.0
4    17450.0
Name: price, dtype: float64

In [18]:
# input
x= cars.drop("price",axis=1)
x.head()

car_ID  symboling  CarName  fueltype  aspiration  doornumber  carbody  \
0       1          3        2         1           0           1        0   
1       2          3        3         1           0           1        0   
2       3          1        1         1           0           1        2   
3       4          2        4         1           0           0        3   
4       5          2        5         1           0           0        3   

   drivewheel  enginelocation  wheelbase  ...  cylindernumber  enginesize  \
0           2               0       88.6  ...               2         130   
1           2               0       88.6  ...               2         130   
2           2               0       94.5  ...               3         152   
3           1               0       99.8  ...               2         109   
4           0               0       99.4  ...               1         136   

   fuelsystem  boreratio  stroke  compressionratio  horsepower  peakrpm  \
0           5       3.47    2.68               9.0         111     5000   
1           5       3.47    2.68               9.0         111     5000   
2           5       2.68    3.47               9.0         154     5000   
3           5       3.19    3.40              10.0         102     5500   
4           5       3.19    3.40               8.0         115     5500   

   citympg  highwaympg  
0       21          27  
1       21          27  
2       19          26  
3       24          30  
4       18          22  

[5 rows x 25 columns]

## Normalizing data / 0 and 1 scaling (Minmax scaling-Bringing all features in the range 0-1)

In [19]:
from sklearn.preprocessing import MinMaxScaler

In [20]:
minMaxScaler= MinMaxScaler(feature_range=(0,1))

In [21]:
# Scaling input column x
# Scaling of output column y is not required since prediction is to be done on this extracted dataset
X = minMaxScaler.fit_transform(x)

In [22]:
X[:5]

array([[0.        , 1.        , 0.01369863, 1.        , 0.        ,
        1.        , 0.        , 1.        , 0.        , 0.05830904,
        0.41343284, 0.31666667, 0.08333333, 0.41117145, 0.        ,
        0.33333333, 0.26037736, 0.71428571, 0.66428571, 0.29047619,
        0.125     , 0.2625    , 0.34693878, 0.22222222, 0.28947368],
       [0.00490196, 1.        , 0.02054795, 1.        , 0.        ,
        1.        , 0.        , 1.        , 0.        , 0.05830904,
        0.41343284, 0.31666667, 0.08333333, 0.41117145, 0.        ,
        0.33333333, 0.26037736, 0.71428571, 0.66428571, 0.29047619,
        0.125     , 0.2625    , 0.34693878, 0.22222222, 0.28947368],
       [0.00980392, 0.6       , 0.00684932, 1.        , 0.        ,
        1.        , 0.5       , 1.        , 0.        , 0.2303207 ,
        0.44925373, 0.43333333, 0.38333333, 0.51784329, 0.83333333,
        0.5       , 0.34339623, 0.71428571, 0.1       , 0.66666667,
        0.125     , 0.44166667, 0.34693878, 0.

In [23]:
y[:5]

0    13495.0
1    16500.0
2    16500.0
3    13950.0
4    17450.0
Name: price, dtype: float64

## Training,Spliting, and Testing the Model

In [24]:
from sklearn.model_selection import train_test_split

In [25]:
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size=0.30,random_state=0)

In [26]:
x_train.shape,y_train.shape

((143, 25), (143,))

In [27]:
x_test.shape,y_test.shape

((62, 25), (62,))

# Technique 1: Linear Regression Model

In [28]:
from sklearn.linear_model import LinearRegression

In [29]:
LinearRegressionModel=LinearRegression()

In [30]:
LinearRegressionModel.fit(x_train,y_train)

LinearRegression()

In [31]:
LinearRegressionModel.score(x_train,y_train)

0.9212207327796392

In [32]:
LinearRegressionModel.score(x_test,y_test)

0.7961110382079569

Since there is a huge difference between accuracy of training and testing data, therefore overfiting of the model takes place.
Hence we will try with some other regression model to overcome this overfiting.

# Technique 2: Decision Tree Regressor

In [33]:
from sklearn.tree import DecisionTreeRegressor

In [34]:
DecisionTreeModel=DecisionTreeRegressor(max_depth=6,random_state=0)

In [35]:
DecisionTreeModel.fit(x_train,y_train)

DecisionTreeRegressor(max_depth=6, random_state=0)

In [36]:
DecisionTreeModel.score(x_train,y_train)

0.9828973841258191

In [37]:
DecisionTreeModel.score(x_test,y_test)

0.8541199129138503

The overfiting of model takes place with Decision Tree Regressor as well, so we will try with some other regression model to overcome this overfiting.

# Technique 3 : Random Forest Regressor

In [38]:
from sklearn.ensemble import RandomForestRegressor

In [39]:
RFModel= RandomForestRegressor(n_estimators=100,max_depth=5,random_state=0)  # n_estimators=100 means number of trees=100

In [40]:
RFModel.fit(x_train,y_train)

RandomForestRegressor(max_depth=5, random_state=0)

In [41]:
RFModel.score(x_train,y_train)

0.9785921788069037

In [42]:
RFModel.score(x_test,y_test)

0.9083065990103958

Now, the model is not overfited and hence a satisfactory accuracy for the training data is obtained that is 90.83%